### PR TITLE
Jan/icp find correspondences

### PR DESCRIPTION
**Pull request by JanRuettinger**
What does the pull request do?
- Find corresponding points in two frames
- Add C++11 in Cmake (hope that's correct)
- Extend Frame class by adding helper functions which are used by the icp class

### CODE REVIEW

File: src/icp/ICP.cpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+// frame Reference Paper:
+// https://www.cvl.iis.u-tokyo.ac.jp/~oishi/Papers/Alignment/Blais_Registering_multiview_PAMI1995.pdf
+// Input: current frame, previous frame, estimated pose of previous frame
+// Output: indices of corresponding vertices in current and previous frame
+// Simple version: only take euclidean distance between points into
+// consideration without normals Advanced version: Euclidean distance between
+// points + difference in normal angles
+std::vector<std::pair<size_t, size_t>> ICP::findIndicesOfCorrespondingPoints(
+    std::shared_ptr<Frame> currFrame, std::shared_ptr<Frame> prevFrame,
+    const Eigen::Matrix4f &estimatedPose) {
```
Comment (VladimirYugay) [inline src/icp/ICP.cpp:69]: What's difference between shared_ptr and directly passing by address?

File: src/icp/ICP.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    if (currPoint.allFinite()) {
+      const Eigen::Vector3f currPointGlobal =
+          rotation * currPoint + translation;
+      // std::cout << "Current Point (Global): " << currPointGlobal[0] << " " <<
+      // currPointGlobal[1] << " " << currPointGlobal[2] << std::endl;
+
+      const Eigen::Vector3f currPointPrevFrame =
+          prevFrame->projectPointIntoFrame(currPointGlobal);
+      const Eigen::Vector2i currPointImgCoordPrevFrame =
+          prevFrame->projectOntoImgPlane(currPointPrevFrame);
```
Comment (VladimirYugay) [inline src/icp/ICP.cpp]: What's the difference between calling a method through "->" vs. ".'? Sometimes you call it via dot and sometimes via an arrow?

File: src/icp/ICP.cpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        // currPointGlobal).norm() << std::endl;
+        if (prevFramePointGlobal.allFinite() &&
+            (prevFramePointGlobal - currPointGlobal).norm() <
+                distanceThreshold) {
+          // std::cout << "Distance: " << (prevFramePointGlobal -
+          // currPointGlobal).norm() << std::endl;
+          indicesOfCorrespondingPoints.push_back(std::make_pair(prevIdx, idx));
+        }
+      } else {
+        // std::cout << "Point not within image" << std::endl;
```
Comment (VladimirYugay) [inline src/icp/ICP.cpp]: These "else" are redundant in case we do not do anything inside

File: src/main.cpp [hunk 4]
```diff
@@ -1,47 +1,58 @@
...
+    Matrix3f depthIntrinsics = sensor.GetDepthIntrinsics();
+    Matrix4f depthExtrinsicsInv = sensor.GetDepthExtrinsics().inverse();
+    Matrix4f trajectory = sensor.GetTrajectory();
+    Matrix4f trajectoryInv = sensor.GetTrajectory().inverse();
+    Matrix3f depthIntrinsicsInv = depthIntrinsics.inverse();
+    int depthHeight = sensor.GetDepthImageHeight();
+    int depthWidth = sensor.GetDepthImageWidth();
+    Matrix4f identity = Matrix4f::Identity(4, 4); // initial estimate
+
+    currFrame = std::make_shared<Frame>(
```
Comment (VladimirYugay) [inline src/main.cpp]: What's the point of making it shared and not just pass by reference?

File: src/models/Frame.cpp [hunk 5]
```diff
@@ -4,56 +4,115 @@
...
 
+Frame::Frame(const float *depthMap, const BYTE *colorMap,
+             const Eigen::Matrix3f &depthIntrinsics,
+             const Eigen::Matrix4f &depthExtrinsicsInv,
+             const Eigen::Matrix4f &trajectoryInv, int depthWidth,
+             int depthHeight) {
+  computeVertexMap(depthMap, depthIntrinsics, depthWidth, depthHeight);
+  computeNormalMap(depthWidth, depthHeight);
+  setExtrinsicMatrix(depthExtrinsicsInv.inverse());
+  this->depthWidth = depthWidth;
```
Comment (VladimirYugay) [inline src/models/Frame.cpp]: "this" is redundant in a constructor, https://en.cppreference.com/w/cpp/language/constructor

File: src/models/Frame.cpp [hunk 6]
```diff
@@ -4,56 +4,115 @@
...
+std::vector<Eigen::Vector3f>
+Frame::transformVertices(const std::vector<Eigen::Vector3f> &points,
+                         const Eigen::Matrix4f &transformation) {
+  const Eigen::Matrix3f rotation = transformation.block(0, 0, 3, 3);
+  const Eigen::Vector3f translation = transformation.block(0, 3, 3, 1);
+  std::vector<Eigen::Vector3f> transformed(points.size());
+
+  for (size_t idx = 0; idx < points.size(); ++idx) {
+    if (points[idx].allFinite())
+      transformed[idx] = rotation * points[idx] + translation;
```
Comment (VladimirYugay) [inline src/models/Frame.cpp:78]: You can init "transformed" with MINF initially and only update it in case of "allFinit()", this will economize operations

File: src/icp/ICP.cpp (hunk 2 above)
Comment (mohamedmerrouch) [reply src/icp/ICP.cpp]: noticed that he uses -> with the "shared_ptr"

File: src/models/Frame.cpp [hunk 7]
```diff
@@ -4,56 +4,115 @@
...
+  return mVerticesGlobal;
+}
+
+Eigen::Vector3f Frame::projectPointIntoFrame(const Eigen::Vector3f &point) {
+  Eigen::Matrix4f extrinsicMatrixInverse = extrinsicMatrix.inverse();
+  // std::cout << extrinsicMatrixInverse << std::endl;
+  const auto rotation = extrinsicMatrixInverse.block(0, 0, 3, 3);
+  const auto translation = extrinsicMatrixInverse.block(0, 3, 3, 1);
+  return rotation * point + translation;
+}
```
Comment (mohamedmerrouch) [inline src/models/Frame.cpp]:
```suggestion
Eigen::Vector3f Frame::projectPointIntoFrame(const Eigen::Vector3f &point) {
  // std::cout << extrinsicMatrixInverse << std::endl;
  const auto rotation = extrinsicMatrix.block(0, 0, 3, 3);
  const auto translation = extrinsicMatrix.block(0, 3, 3, 1);
  return rotation * point + translation;
}
```
Ths helper function is used to transform point from global coordinate to frame coordinate. I guess extrinsic matrix is the one to be used and not its inverse.

File: src/icp/ICP.cpp (hunk 1 above)
Comment (VladimirYugay) [reply src/icp/ICP.cpp:69]: I also think, that "curFrame" and "prevFrame" should be passed by a constant reference to avoid possible bugs in the future

File: src/icp/ICP.cpp (hunk 2 above)
Comment (JanRuettinger) [reply src/icp/ICP.cpp]: Here is a [SO thread](https://stackoverflow.com/questions/33901858/how-does-unique-ptr-support-both-dot-and-arrow-invocation-and-undefined-met ) about the topic. 
tldr; smart pointer overload the "->" operator, so that when we call `smartpointer->method` we access the method of the object the smartpointer points to. If we do `smartpointer.method` we access the method of the smartpointer object which is not what we want. I would expect the dot version to result in an error.

File: src/main.cpp (hunk 4 above)
Comment (JanRuettinger) [reply src/main.cpp]: I tried using a reference before but ran into the following issue:
```
Frame* prevFrame; // no value assigned in first run
Frame* curFrame;
while(..){

curFrame = Frame(...)

if (we have at least two frames){
  findIndicesOfCorrespondingPoints(currFrame, prevFrame, ...);
}
prevFrame = curFrame // needs to stay in memory for next run of while loop
}
```
I guess the easiest option is to use pointers for this problem. Since you should avoid using raw points, a shared pointer was the next best option. If we don't want to use pointers here, then we need an empty constructor to create an empty prevFrame Frame object above the while loop and a move/copy constructor for the assignment at the end.

File: src/models/Frame.cpp (hunk 5 above)
Comment (JanRuettinger) [reply src/models/Frame.cpp]: Hm I feel like by using this the code is more explicit. I guess using it doesn't harm us in any way.

File: src/models/Frame.cpp (hunk 7 above)
Comment (JanRuettinger) [reply src/models/Frame.cpp]: Yes, I agree.

File: CMakeLists.txt [hunk 8]
```diff
@@ -1,6 +1,8 @@
 cmake_minimum_required(VERSION 3.6)
```
Comment (VladimirYugay) [inline CMakeLists.txt:1]: Can you squash your commits? There are 11 and some of them are pointing to a broken code. If somebdy will try to rebase that would be an intense pain for them to follow 11 commits.

You can easily do this without any additional work by rebasing on one of your own commits in the same branch.

Example:
git rebase -i a047633933ee594f30550c40c595d88c72c043c6

After that, in the opened window, instead of "pick" in front of each commit message, you write "f" which means "fixup".

This will literally "remove and paste" your commits inside your own code. There will be no conflicts and other unpleasant things

**Review by mohamedmerrouch — APPROVED** [review]
👌

File: CMakeLists.txt (hunk 8 above)
Comment (JanRuettinger) [reply CMakeLists.txt:1]: Yes makes sense, thanks!

What would you say is best practice? 
1. Work on your feature branch commit often and then do a PR and start the review discussion?
=> Potentially you can go back 5 commits to do more changes

2. Squash all commits and only then create the PR
=> More difficult to revert single commits later

3. Hybrid approach: commit often but rewrite your history from time to time to make it easier for other people to understand

Comment (VladimirYugay) [reply CMakeLists.txt:1]: Cool. 

I'm not really an expert in this thing, just the practices from work.

In theory there's smth called [gitflow](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow), but for us it's not that strictly necessary.

In general, no commit should break the tests (in our case no commit should break the code), since somebody might checkout it and be in a complete shit. His new features would be based on erroneous code, he will try to fix them and then when time for rebasing comes everybody is super angry. 

However, there're cases when you add smal fixes like linting or some other stuff to pass PR review. After review is finished and all your changes approved, you usually "squash" your small unimprotant commits like you did.
It's not really squash, rather rebase, there's a `git squash` for it, but rebasing is somewhat easier I guess.

**Review by VladimirYugay — APPROVED** [review]
